File: backend/old/Backend.cpp
#include <iostream>
#include <thread>
#include <vector>
#include <string>
#include <cstring>
#include <unistd.h>
#include <netinet/in.h>
#include <nlohmann/json.hpp>

using json = nlohmann::json;

#define PORT 65433
#define BUFFER_SIZE 1024

void handle_client(int client_socket) {
    char buffer[BUFFER_SIZE];
    std::string mode = "";

    while (true) {
        memset(buffer, 0, BUFFER_SIZE);
        ssize_t bytes_received = recv(client_socket, buffer, BUFFER_SIZE, 0);
        if (bytes_received <= 0) {
            break; // Client disconnected
        }

        try {
            json data = json::parse(buffer);
            std::cout << "Reçu du client : " << data.dump() << std::endl;

            if (data.contains("mode"))
                mode = data["mode"];
            else
            {
                // [human vs ai] & [human vs human] : coup autorisé ?

                int x = data["x"];
                int y = data["y"];

                json json_game_state = json::array();

                json_game_state.push_back({
                    {"x", x},
                    {"y", y},
                    {"color", "black"}
                });

                json response = {
                    {"game_state", json_game_state},
                    {"win", false},
                    {"authorized", true},
                };

                std::string response_str = response.dump();
                send(client_socket, response_str.c_str(), response_str.length(), 0);
            }
        } catch (const std::exception& e) {
            std::cerr << "Erreur JSON: " << e.what() << std::endl;
        }
    }

    close(client_socket);
}

void run_server() {
    int server_fd = socket(AF_INET, SOCK_STREAM, 0);
    if (server_fd == 0) {
        perror("socket failed");
        exit(EXIT_FAILURE);
    }

    sockaddr_in address;
    socklen_t addrlen = sizeof(address);
    address.sin_family = AF_INET;
    address.sin_addr.s_addr = INADDR_ANY; // localhost
    address.sin_port = htons(PORT);

    if (bind(server_fd, (struct sockaddr*)&address, sizeof(address)) < 0) {
        perror("bind failed");
        exit(EXIT_FAILURE);
    }

    if (listen(server_fd, 3) < 0) {
        perror("listen");
        exit(EXIT_FAILURE);
    }

    std::cout << "Serveur IA en écoute sur le port " << PORT << std::endl;

    while (true) {
        int client_socket = accept(server_fd, (struct sockaddr*)&address, &addrlen);
        if (client_socket < 0) {
            perror("accept");
            continue;
        }
        handle_client(client_socket);
    }
}

int main() {
    run_server();
    return 0;
}

File: backend/old/backend.py
import socket
import threading
import json

HOST = 'localhost'
PORT = 65433


def handle_client(conn):
    mode = None
    game_state = {}
    current_player = "black"
    
    with conn:
        while True:
            data = received_data(conn)

            if "mode" in data:
                mode = chosen_mode(data)

            else:
                # [human vs ai] & [human vs human] : coup autorisé ?
                game_state, authorized = rules(data, game_state, current_player)

                # reaction en fonction du mode
                if mode == "human":
                    current_player = "white" if current_player == "black" else "black"
                if mode == "ai":
                    game_state = ai(data, game_state, current_player)

                # Conversion pour envoie de l'état de jeu
                json_game_state = [{"x": x, "y": y, "color": c} for (x, y), c in game_state.items()]
                response = {
                    "game_state" : json_game_state,
                    "win" : win(data),
                    "authorized": authorized
                    }
                # Envoie au front
                conn.sendall(json.dumps(response).encode())


def rules(data, game_state, current_player):
    BOARD_SIZE = 19
    x, y = data['x'], data['y']
    authorized = True

    if (x, y) in game_state or not (0 <= x < BOARD_SIZE and 0 <= y < BOARD_SIZE):
        authorized = False

    if authorized:
        game_state[(x, y)] = current_player
    return game_state, authorized


def win(data):
    return False


def ai(data, game_state, current_player): # last change 23.08
    x, y = data['x'], data['y']
    current_player = "white" if current_player == "black" else "black"

    game_state[(x - 1, y)] = current_player
    game_state.pop((x + 1, y), None)

    return game_state


def received_data(conn):
    data = conn.recv(1024)
    if not data:
        return
    data = json.loads(data.decode())
    # print("Reçu du client :", data)
    return data


def chosen_mode(data):
    if "mode" in data:
        print(f"Mode défini pour ce client : {data['mode']}")
        return data["mode"]


def run_server():
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
        s.bind((HOST, PORT))
        s.listen()
        print(f"Serveur en écoute sur {HOST}:{PORT}")
        while True:
            conn, _ = s.accept()
            threading.Thread(target=handle_client, args=(conn,), daemon=True).start()


if __name__ == "__main__":
    run_server()
